File: android/CodexFlowAndroid/app/src/main/java/com/codexflow/codexflow/data/model/SessionModels.kt
package com.codexflow.codexflow.data.model

import java.time.Instant
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.util.Locale
import kotlinx.serialization.Serializable

@Serializable
data class SessionListEnvelope(
    val data: List<SessionSummary> = emptyList()
)

@Serializable
data class SessionSummary(
    val id: String = "",
    val name: String = "",
    val preview: String = "",
    val cwd: String = "",
    val source: String = "",
    val status: String = "",
    val activeFlags: List<String> = emptyList(),
    val loaded: Boolean = false,
    val updatedAt: Long = 0,
    val createdAt: Long = 0,
    val modelProvider: String = "",
    val branch: String = "",
    val pendingApprovals: Int = 0,
    val lastTurnId: String = "",
    val lastTurnStatus: String = "",
    val agentNickname: String = "",
    val agentRole: String = "",
    val ended: Boolean = false
) {
    val displayName: String
        get() = normalizedTitle(name)
            ?: normalizedTitle(agentNickname)
            ?: cwd.trim().substringAfterLast('/').takeIf { it.isNotBlank() }
            ?: previewSummary.takeIf { it.isNotBlank() }?.take(32)
            ?: "Session ${id.take(8)}"

    val previewSummary: String
        get() = preview
            .lineSequence()
            .map { it.trim().replace("\t", " ") }
            .firstOrNull { it.isNotBlank() }
            .orEmpty()

    val previewExcerpt: String
        get() = preview
            .splitToSequence(Regex("\\s+"))
            .filter { it.isNotBlank() }
            .joinToString(" ")
            .headTailTruncated()

    val updatedAtDisplay: String
        get() = formatEpochSeconds(updatedAt)

    val hasWaitingState: Boolean
        get() = activeFlags.contains("waitingOnApproval") || activeFlags.contains("waitingOnUserInput")

    val isActive: Boolean
        get() = status == "active"

    val isRunningTurn: Boolean
        get() = lastTurnStatus == "inProgress"

    private fun normalizedTitle(value: String): String? = value.trim().takeIf { it.isNotEmpty() }
}

@Serializable
data class SessionDetail(
    val summary: SessionSummary = SessionSummary(),
    val turns: List<TurnDetail> = emptyList()
)

@Serializable
data class TurnDetail(
    val id: String = "",
    val status: String = "",
    val startedAt: Long = 0,
    val completedAt: Long = 0,
    val durationMs: Long = 0,
    val error: String = "",
    val diff: String = "",
    val planExplanation: String = "",
    val plan: List<PlanStep> = emptyList(),
    val items: List<TurnItem> = emptyList()
)

@Serializable
data class PlanStep(
    val step: String = "",
    val status: String = ""
)

@Serializable
data class TurnItem(
    val id: String = "",
    val type: String = "",
    val title: String = "",
    val body: String = "",
    val status: String = "",
    val auxiliary: String = "",
    val metadata: Map<String, String> = emptyMap()
)

fun String.headTailTruncated(maxLength: Int = 220, head: Int = 140, tail: Int = 72): String {
    val trimmed = trim()
    if (trimmed.length <= maxLength) return trimmed
    if (head <= 0 || tail <= 0 || head + tail >= trimmed.length) return trimmed
    return trimmed.take(head) + " ... " + trimmed.takeLast(tail)
}

fun formatEpochSeconds(timestamp: Long): String {
    if (timestamp <= 0) return "未知"
    val dateTime = Instant.ofEpochSecond(timestamp).atZone(ZoneId.systemDefault())
    val now = Instant.now().atZone(ZoneId.systemDefault())
    val formatter = when {
        dateTime.toLocalDate() == now.toLocalDate() ->
            DateTimeFormatter.ofPattern("今天 HH:mm", Locale.CHINA)
        dateTime.toLocalDate() == now.toLocalDate().minusDays(1) ->
            DateTimeFormatter.ofPattern("昨天 HH:mm", Locale.CHINA)
        dateTime.year == now.year ->
            DateTimeFormatter.ofPattern("MM-dd HH:mm", Locale.CHINA)
        else ->
            DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm", Locale.CHINA)
    }
    return dateTime.format(formatter)
}

File: windows/CodexFlowWindows/src/CodexFlowWindows/Data/Models/SessionModels.cs
using System.Globalization;
using System.Text.Json.Serialization;

namespace CodexFlowWindows.Data.Models;

public sealed record SessionListEnvelope
{
    [JsonPropertyName("data")]
    public IReadOnlyList<SessionSummary> Data { get; init; } = Array.Empty<SessionSummary>();
}

public sealed record SessionSummary
{
    [JsonPropertyName("id")]
    public string Id { get; init; } = string.Empty;

    [JsonPropertyName("name")]
    public string Name { get; init; } = string.Empty;

    [JsonPropertyName("preview")]
    public string Preview { get; init; } = string.Empty;

    [JsonPropertyName("cwd")]
    public string Cwd { get; init; } = string.Empty;

    [JsonPropertyName("source")]
    public string Source { get; init; } = string.Empty;

    [JsonPropertyName("status")]
    public string Status { get; init; } = string.Empty;

    [JsonPropertyName("activeFlags")]
    public IReadOnlyList<string> ActiveFlags { get; init; } = Array.Empty<string>();

    [JsonPropertyName("loaded")]
    public bool Loaded { get; init; }

    [JsonPropertyName("updatedAt")]
    public long UpdatedAt { get; init; }

    [JsonPropertyName("createdAt")]
    public long CreatedAt { get; init; }

    [JsonPropertyName("modelProvider")]
    public string ModelProvider { get; init; } = string.Empty;

    [JsonPropertyName("branch")]
    public string Branch { get; init; } = string.Empty;

    [JsonPropertyName("pendingApprovals")]
    public int PendingApprovals { get; init; }

    [JsonPropertyName("lastTurnId")]
    public string LastTurnId { get; init; } = string.Empty;

    [JsonPropertyName("lastTurnStatus")]
    public string LastTurnStatus { get; init; } = string.Empty;

    [JsonPropertyName("agentNickname")]
    public string AgentNickname { get; init; } = string.Empty;

    [JsonPropertyName("agentRole")]
    public string AgentRole { get; init; } = string.Empty;

    [JsonPropertyName("ended")]
    public bool Ended { get; init; }

    [JsonIgnore]
    public string DisplayName =>
        FirstNonEmpty(Name, AgentNickname, DirectoryName, PreviewSummary) is { Length: > 0 } value
            ? Truncate(value, 48)
            : $"Session {ShortId}";

    [JsonIgnore]
    public string PreviewSummary => Preview
        .Replace('\t', ' ')
        .Split(new[] { '\r', '\n' }, StringSplitOptions.RemoveEmptyEntries)
        .Select(line => line.Trim())
        .FirstOrDefault(line => line.Length > 0) ?? string.Empty;

    [JsonIgnore]
    public string PreviewExcerpt => HeadTailTruncated(
        string.Join(" ", Preview.Split((char[]?)null, StringSplitOptions.RemoveEmptyEntries)),
        220,
        140,
        72);

    [JsonIgnore]
    public string UpdatedAtDisplay => FormatEpochSeconds(UpdatedAt);

    [JsonIgnore]
    public bool HasWaitingState =>
        ActiveFlags.Contains("waitingOnApproval") || ActiveFlags.Contains("waitingOnUserInput") || PendingApprovals > 0;

    [JsonIgnore]
    public bool IsRunningTurn => LastTurnStatus == "inProgress";

    [JsonIgnore]
    public string ShortId => Id.Length <= 8 ? Id : Id[..8];

    [JsonIgnore]
    private string DirectoryName
    {
        get
        {
            var trimmed = Cwd.Trim();
            if (trimmed.Length == 0)
            {
                return string.Empty;
            }
            return trimmed.TrimEnd('\\', '/').Split('\\', '/').LastOrDefault() ?? string.Empty;
        }
    }

    private static string? FirstNonEmpty(params string[] values) =>
        values.FirstOrDefault(value => !string.IsNullOrWhiteSpace(value))?.Trim();

    private static string Truncate(string value, int maxLength) =>
        value.Length <= maxLength ? value : value[..maxLength] + "...";

    private static string HeadTailTruncated(string value, int maxLength, int head, int tail)
    {
        var trimmed = value.Trim();
        if (trimmed.Length <= maxLength || head <= 0 || tail <= 0 || head + tail >= trimmed.Length)
        {
            return trimmed;
        }
        return trimmed[..head] + " ... " + trimmed[^tail..];
    }

    private static string FormatEpochSeconds(long timestamp)
    {
        if (timestamp <= 0)
        {
            return "未知";
        }

        var dateTime = DateTimeOffset.FromUnixTimeSeconds(timestamp).ToLocalTime();
        var now = DateTimeOffset.Now;
        if (dateTime.Date == now.Date)
        {
            return "今天 " + dateTime.ToString("HH:mm", CultureInfo.GetCultureInfo("zh-CN"));
        }
        if (dateTime.Date == now.Date.AddDays(-1))
        {
            return "昨天 " + dateTime.ToString("HH:mm", CultureInfo.GetCultureInfo("zh-CN"));
        }
        if (dateTime.Year == now.Year)
        {
            return dateTime.ToString("MM-dd HH:mm", CultureInfo.GetCultureInfo("zh-CN"));
        }
        return dateTime.ToString("yyyy-MM-dd HH:mm", CultureInfo.GetCultureInfo("zh-CN"));
    }
}

public sealed record SessionDetail
{
    [JsonPropertyName("summary")]
    public SessionSummary Summary { get; init; } = new();

    [JsonPropertyName("turns")]
    public IReadOnlyList<TurnDetail> Turns { get; init; } = Array.Empty<TurnDetail>();
}

public sealed record TurnDetail
{
    [JsonPropertyName("id")]
    public string Id { get; init; } = string.Empty;

    [JsonPropertyName("status")]
    public string Status { get; init; } = string.Empty;

    [JsonPropertyName("startedAt")]
    public long StartedAt { get; init; }

    [JsonPropertyName("completedAt")]
    public long CompletedAt { get; init; }

    [JsonPropertyName("durationMs")]
    public long DurationMs { get; init; }

    [JsonPropertyName("error")]
    public string Error { get; init; } = string.Empty;

    [JsonPropertyName("diff")]
    public string Diff { get; init; } = string.Empty;

    [JsonPropertyName("planExplanation")]
    public string PlanExplanation { get; init; } = string.Empty;

    [JsonPropertyName("plan")]
    public IReadOnlyList<PlanStep> Plan { get; init; } = Array.Empty<PlanStep>();

    [JsonPropertyName("items")]
    public IReadOnlyList<TurnItem> Items { get; init; } = Array.Empty<TurnItem>();

    [JsonIgnore]
    public string DurationDisplay => DurationMs > 0 ? $"{DurationMs / 1000.0:0.0}s" : string.Empty;
}

public sealed record PlanStep
{
    [JsonPropertyName("step")]
    public string Step { get; init; } = string.Empty;

    [JsonPropertyName("status")]
    public string Status { get; init; } = string.Empty;
}

public sealed record TurnItem
{
    [JsonPropertyName("id")]
    public string Id { get; init; } = string.Empty;

    [JsonPropertyName("type")]
    public string Type { get; init; } = string.Empty;

    [JsonPropertyName("title")]
    public string Title { get; init; } = string.Empty;

    [JsonPropertyName("body")]
    public string Body { get; init; } = string.Empty;

    [JsonPropertyName("status")]
    public string Status { get; init; } = string.Empty;

    [JsonPropertyName("auxiliary")]
    public string Auxiliary { get; init; } = string.Empty;

    [JsonPropertyName("metadata")]
    public IReadOnlyDictionary<string, string> Metadata { get; init; } = new Dictionary<string, string>();
}
